src/decoder.rs

```diff
@@ -0,0 +1,77 @@
+use DecodeError;
+
+pub struct AsciiDecoder;
+pub struct Utf8Decoder;
+
+macro_rules! decode {
+    ($alpha:ident, $input:ident, $iter:ident, $c:pat => $carry:expr) => ({
+        if $input.len() == 0 {
+            return Ok(Vec::new());
+        }
+
+        let base = $alpha.len() as u16;
+
+        let mut bytes = Vec::with_capacity($input.len());
+        bytes.push(0u8);
+
+        for $c in $input.$iter() {
+            let mut carry = $carry as u16;
+
+            for byte in bytes.iter_mut() {
+                carry += base * *byte as u16;
+                *byte = carry as u8;
+                carry >>= 8;
+            }
+
+            while carry > 0 {
+                bytes.push(carry as u8);
+                carry >>= 8;
+            }
+        }
+
+        let leader = $alpha[0];
+
+        let leaders = $input
```

Comment:
Looking at the code in https://github.com/bitcoin/bitcoin/blob/master/src/base58.cpp it seems to me that there is an optimization we are missing, which is doing the leader handling before the actual encode decode handling, such that, those can be actually skipped from the main iteration. 

There is also strim trimming happening in there which might make sense to add in here as well.